### ISSUE TITLE
PERF: concat along axis 1 unnecessarily materializes RangeIndex->Int64Index

### ISSUE THREAD

**Issue by vyasr** [pandas-dev/pandas#47501]
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this issue exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this issue exists on the main branch of pandas.


### Reproducible Example

This is as of pandas 1.4.3
```
>>> import pandas as pd                                                                                                                                                                                                                                                                                                                                                
>>> pd.concat([pd.DataFrame({'a': range(10)}), pd.DataFrame({'b': range(10)})], sort=True, axis=1).index                                                                                                   Int64Index([0, 1, 2, 3, 4, 5, 6, 7, 8, 9], dtype='int64')
```

Even though both inputs have identical `RangeIndex` inputs, the output index is an Int64Index. This behavior appears to be triggered by the `sort=True` parameter, since removing that gives a `RangeIndex`.

```
>>> pd.concat([pd.DataFrame({'a': range(10)}), pd.DataFrame({'b': range(10)})], axis=1).index
RangeIndex(start=0, stop=10, step=1)
```

My naive guess is that there is a missing check somewhere that the sort is a no-op on a `RangeIndex` .

This issue definitely seems related to https://github.com/pandas-dev/pandas/issues/46675, but it is not identical since it has specifically appeared in 1.4.3, whereas that issue was already present in 1.4.2.

### Installed Versions

<details>
```
commit           : e8093ba372f9adfe79439d90fe74b0b5b6dea9d6
python           : 3.8.13.final.0
python-bits      : 64
OS               : Linux
OS-release       : 4.15.0-76-generic
Version          : #86-Ubuntu SMP Fri Jan 17 17:24:28 UTC 2020
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : en_US.UTF-8
pandas           : 1.4.3
numpy            : 1.22.4
pytz             : 2022.1
dateutil         : 2.8.2
setuptools       : 62.6.0
pip              : 22.1.2
Cython           : 0.29.30
pytest           : 7.1.2
hypothesis       : 6.47.1
sphinx           : 5.0.2
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.9.0
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 3.0.3
IPython          : 8.4.0
pandas_datareader: None
bs4              : 4.11.1
bottleneck       : None
brotli           :
fastparquet      : None
fsspec           : 2022.5.0
gcsfs            : None
markupsafe       : 2.1.1
matplotlib       : 3.5.2
numba            : 0.55.2
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : 8.0.0
pyreadstat       : None
pyxlsb           : None
s3fs             : 2022.5.0
scipy            : 1.8.1
snappy           :
sqlalchemy       : 1.4.38
tables           : None
tabulate         : 0.8.10
xarray           : None
xlrd             : None
xlwt             : None
zstandard        : None
```
</details>


### Prior Performance

This is as of pandas 1.4.2
```
>>> pd.concat([pd.DataFrame({'a': range(10)}), pd.DataFrame({'b': range(10)})], sort=True, axis=1).index
RangeIndex(start=0, stop=10, step=1)
```

**Comment by jreback** [comment]
can u show how much this actually matters at scale eg 100 columns 100k rows and if it matters

**Comment by simonjayhawkins** [comment]
> This issue definitely seems related to #46675, but it is not identical since it has specifically appeared in 1.4.3, whereas that issue was already present in 1.4.2.

first bad commit: [75a799cbf108deae6228ba4ab00c0ff19008bba6] Backport PR #47206 on branch 1.4.x (REGR: concat not sorting columns for mixed column names) (#47251)

@phofl it appears that #47508 does not fix #46675, are they not related?

**Comment by phofl** [comment]
No I don't think so. The other is related to the use of merge under the hood I think
